Erroneous "Cannot infer type arguments" error

stephan-herrmann opened the pull request:
fixes #1475

Implementation of JLS 15.9.3. failed to substitute an enclosing type in
>  - The return type of mj is θj applied to D<F1, ..., Fp>.

srikanth-sankaran reviewed (approved):
Looks good. Mmm what has taken so long for this to be caught ?

srikanth-sankaran commented:
A quick pass over some of the existing bugs: 

https://bugs.eclipse.org/bugs/show_bug.cgi?id=576846 - not reproducible with or without this fix - should be closed with a regression test.
https://bugs.eclipse.org/bugs/show_bug.cgi?id=569231 - reproducible with or without this fix
https://bugs.eclipse.org/bugs/show_bug.cgi?id=568586 - Lacks a full test case
https://bugs.eclipse.org/bugs/show_bug.cgi?id=566989 - not reproducible with or without this fix - should be closed with a regression test.
https://bugs.eclipse.org/bugs/show_bug.cgi?id=551929 - reproducible with or without this fix
https://bugs.eclipse.org/bugs/show_bug.cgi?id=509848 - not reproducible with or without this fix - should be closed with a regression test.

There could be value in coming up with regression tests from the working cases ((after comparing behavior with javac  of course) - even if they were not fixed by this PR - so we have our test suite running them on a regular basis

stephan-herrmann commented:
Some bisecting:

* https://bugs.eclipse.org/bugs/show_bug.cgi?id=569231
  * fixed by #1557 
* https://bugs.eclipse.org/bugs/show_bug.cgi?id=566989
  * also :smile:  fixed by #1557 
* https://bugs.eclipse.org/bugs/show_bug.cgi?id=509848
  * fixed since ecj-4.11 (v20190306-2240) - too old for my git-bisect environment :)

regression tests coming in a minute

stephan-herrmann commented:
> Mmm what has taken so long for this to be caught ?

Java provides too many features for all Java programmers on the planet to ever try all feature combinations? :smile: 

To trigger the bug you need:
* a generic member type
* hosted in a generic enclosing type
* a generic method of that enclosing type
* inside that method a diamond instantiation of the member type

jarthana commented:
We seem to be rejecting the following code (which Javac accepts) after this fix:
```
class Outer <T extends Number>{
    public void foo(){
        class ABC <U extends Number> {
            public ABC(int i, long l, String s, U t) {}
        }
        ABC<?> a = new ABC<>(5, 3L, "", 5) {};
    }
}
```
